### PR TITLE
[data] link dataset ids in constructor, return correct metrics id for materialize

### PR DESCRIPTION
**Pull request by Zandew**
## Why are these changes needed?

The dataset uuid that we record metrics with and the dataset uuid we return from `Dataset.materialize()` are different currently for two reasons:

`ExecutionPlan` is not getting it's dataset uuid as a parameter. The`dataset_uuid` parameter is only set in the deepcopy for `Dataset.serialize_lineage`. It's not useful anywhere else because we need to create the `ExecutionPlan` before the `Dataset`, and pass the plan into the dataset's constructor. Instead, we can link the dataset ids in the dataset's constructor. 

In `materialize`, we create a copy of the current dataset and execute it (this is the uuid that is used for metrics). We then create an output dataset (different uuid) and return that. This pr copies over the uuid of `copy` to `output`. Only `output` gets returned, and `copy` gets destroyed so this shouldn't create 2 datasets with same id. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/plan.py [hunk 1]
```diff
@@ -135,9 +133,7 @@ def __init__(
         self._last_optimized_stages = None
         # Cached schema.
         self._schema = None
-        self._dataset_uuid = dataset_uuid or uuid.uuid4().hex
-        if not stats.dataset_uuid:
-            stats.dataset_uuid = self._dataset_uuid
+        self._dataset_uuid = None
```
Comment (scottjlee) [inline python/ray/data/_internal/plan.py:136]: could we leave a comment here briefly explaining how this is set?

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -243,6 +243,7 @@ def __init__(
 
         self._plan = plan
         self._uuid = uuid4().hex
+        self._plan._dataset_uuid = self._uuid
```
Comment (scottjlee) [inline python/ray/data/dataset.py:246]: we could clean this up with `self._set_uuid(uuid4().hex)`

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -4557,6 +4558,9 @@ def materialize(self) -> "MaterializedDataset":
         )
         output._plan.execute()  # No-op that marks the plan as fully executed.
         output._plan._in_stats.dataset_uuid = self._get_uuid()
```
Comment (scottjlee) [inline python/ray/data/dataset.py:4560]: do we need to update this as well?
`output._plan._in_stats.dataset_uuid = copy._get_uuid()`
also, should we instead use `output._set_uuid(<self/copy>._get_uuid())` for this and the following line to set the uuid of the underlying plan?

Comment (Zandew) [reply python/ray/data/dataset.py:4560]: yeah, I think we should be updating that as well. Also I think we can move the update of `output._plan._in_stats.dataset_uuid` into `Dataset._set_uuid` too.

**Review by scottjlee — APPROVED** [review]

**Review by raulchen — APPROVED** [review]

**Pull request merged**